### PR TITLE
[Vmware][Deploy-as-is] Refactor the OVF parsing

### PR DESCRIPTION
**Pull request by nvazquez**
### Description

This PR refactors the OVF parsing

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### How Has This Been Tested?
- Register OVA templates including templates containing prefixes
- Check each template registration is successful and deploy from them

### CODE REVIEW

File: api/src/main/java/com/cloud/agent/api/storage/OVFHelper.java [hunk 1]
```diff
@@ -366,21 +387,25 @@ private DatadiskTO generateDiskTO(OVFFile file, OVFDisk disk, String ovfParentPa
...
+            if (nodeList.getLength() > 0) {
+                break;
+            }
+        }
+        return nodeList;
+    }
 
-        int totalDisksLength = disks.getLength() + ovfDisks.getLength();
+    protected List<OVFDisk> extractDisksFromOvfDocumentTree(Document doc, List<String> supportedNamespaces) {
+        NodeList disks = getElementByTagNameAndSupportedNamespaces(doc, supportedNamespaces, "Disk");
```
Comment (DaanHoogland) [inline api/src/main/java/com/cloud/agent/api/storage/OVFHelper.java:402]: this will work in most cases. the only quirk is if there are two name spaces defined that both have a `Disk` element.
Can we specify the exact name space we need?

File: api/src/main/java/com/cloud/agent/api/storage/OVFHelper.java [hunk 2]
```diff
@@ -366,21 +387,25 @@ private DatadiskTO generateDiskTO(OVFFile file, OVFDisk disk, String ovfParentPa
...
+        NodeList items = getElementByTagNameAndSupportedNamespaces(doc, supportedNamespaces, "Item");
```
Comment (sureshanaparti) [inline api/src/main/java/com/cloud/agent/api/storage/OVFHelper.java:403]: @nvazquez Any possibility of different namespaces for Disk and Item?

File: api/src/main/java/com/cloud/agent/api/storage/OVFHelper.java (hunk 1 above)
Comment (nvazquez) [reply api/src/main/java/com/cloud/agent/api/storage/OVFHelper.java:402]: Yes we could, but in that case which namespace should be used? If both `Disk` elements have valid prefixes/namespaces then I think the method should retrieve all the elements defined in the file even with different namespaces, do you agree?

File: api/src/main/java/com/cloud/agent/api/storage/OVFHelper.java (hunk 2 above)
Comment (nvazquez) [reply api/src/main/java/com/cloud/agent/api/storage/OVFHelper.java:403]: Yes, it is possible. This may be related to @DaanHoogland's question above ^^

File: api/src/main/java/com/cloud/agent/api/storage/OVFHelper.java (hunk 1 above)
Comment (DaanHoogland) [reply api/src/main/java/com/cloud/agent/api/storage/OVFHelper.java:402]: No, they would have different meaning. Only the one that is meant to be in the name space that usually uses `ovf`as prefix is an actual disk template. I am not to worried about this occurring with high probability but it is a real option for xml-composers.

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test centos7 vmware-67u3

File: api/src/main/java/com/cloud/agent/api/storage/OVFParser.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    /**
+     * Get the text value of a node's child with name or suffix "childNodeName", null if not present
+     * Example:
+     * <Node>
+     *    <childNodeName>Text value</childNodeName>
+     *    <rasd:childNodeName>Text value</rasd:childNodeName>
+     * </Node>
+     */
+    public String getChildNodeValue(Node node, String childNodeName) {
+        if (node != null && node.hasChildNodes()) {
```
Comment (GutoVeronezi) [inline api/src/main/java/com/cloud/agent/api/storage/OVFParser.java:98]: We could invert this `if` statement to reduce indentation.

**Review by GutoVeronezi — COMMENTED** [review]
Nice work, @nvazquez!

File: api/src/main/java/com/cloud/agent/api/storage/OVFParser.java (hunk 3 above)
Comment (nvazquez) [reply api/src/main/java/com/cloud/agent/api/storage/OVFParser.java]: Thanks @GutoVeronezi, done

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test centos7 vmware-67u3

File: api/src/main/java/com/cloud/agent/api/storage/OVFHelper.java [hunk 4]
```diff
@@ -208,11 +148,11 @@ protected OVFPropertyTO createOVFPropertyFromNode(Node node, int index, String c
                         continue;
                     }
                     if (child.getNodeName().equalsIgnoreCase("Category") ||
-                            child.getNodeName().equalsIgnoreCase("ovf:Category")) {
+                            child.getNodeName().endsWith(":Category")) {
```
Comment (DaanHoogland) [inline api/src/main/java/com/cloud/agent/api/storage/OVFHelper.java:151]: bit of a short cut, but it'll work

File: api/src/test/java/com/cloud/agent/api/storage/OVFHelperTest.java [hunk 5]
```diff
@@ -27,13 +27,13 @@
...
 import org.xml.sax.SAXParseException;
 
-import javax.xml.parsers.ParserConfigurationException;
 import java.io.IOException;
 import java.util.List;
 
 public class OVFHelperTest {
 
     private String ovfFileProductSection =
+        "<Envelope xmlns=\"http://schemas.dmtf.org/ovf/envelope/1\" xmlns:cim=\"http://schemas.dmtf.org/wbem/wscim/1/common\" xmlns:ovf=\"http://schemas.dmtf.org/ovf/envelope/1\" xmlns:rasd=\"http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_ResourceAllocationSettingData\" xmlns:vmw=\"http://www.vmware.com/schema/ovf\" xmlns:vssd=\"http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_VirtualSystemSettingData\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\">" +
```
Comment (DaanHoogland) [inline api/src/test/java/com/cloud/agent/api/storage/OVFHelperTest.java:36]: you could put those schema definitions on the `<ProductSection>` itself.

File: api/src/test/java/com/cloud/agent/api/storage/OVFHelperTest.java [hunk 6]
```diff
@@ -44,9 +44,11 @@
                 "<Label>User data to be made available inside the instance</Label>" +
                 "<Description>This allows to pass any text to the appliance. The value should be encoded in base64</Description>" +
             "</Property>" +
-        "</ProductSection>";
+        "</ProductSection>" +
+        "</Envelope>";
 
     private String ovfFileDeploymentOptionsSection =
+            "<Envelope xmlns=\"http://schemas.dmtf.org/ovf/envelope/1\" xmlns:cim=\"http://schemas.dmtf.org/wbem/wscim/1/common\" xmlns:ovf=\"http://schemas.dmtf.org/ovf/envelope/1\" xmlns:rasd=\"http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_ResourceAllocationSettingData\" xmlns:vmw=\"http://www.vmware.com/schema/ovf\" xmlns:vssd=\"http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_VirtualSystemSettingData\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\">" +
```
Comment (DaanHoogland) [inline api/src/test/java/com/cloud/agent/api/storage/OVFHelperTest.java:51]: same here

File: api/src/test/java/com/cloud/agent/api/storage/OVFHelperTest.java [hunk 7]
```diff
@@ -61,9 +63,11 @@
                     "      <Label>2 Gbps (ASAv30)</Label>\n" +
                     "      <Description>Use this option to deploy an ASAv with a maximum throughput of 2 Gbps (uses 4 vCPUs and 8 GB of memory).</Description>\n" +
                     "    </Configuration>\n" +
-                    "  </DeploymentOptionSection>";
+                    "  </DeploymentOptionSection>" +
+            "</Envelope>";
 
     private String ovfFileVirtualHardwareSection =
+            "<Envelope xmlns=\"http://schemas.dmtf.org/ovf/envelope/1\" xmlns:cim=\"http://schemas.dmtf.org/wbem/wscim/1/common\" xmlns:ovf=\"http://schemas.dmtf.org/ovf/envelope/1\" xmlns:rasd=\"http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_ResourceAllocationSettingData\" xmlns:vmw=\"http://www.vmware.com/schema/ovf\" xmlns:vssd=\"http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_VirtualSystemSettingData\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\">" +
```
Comment (DaanHoogland) [inline api/src/test/java/com/cloud/agent/api/storage/OVFHelperTest.java:70]: and here

File: api/src/test/java/com/cloud/agent/api/storage/OVFHelperTest.java [hunk 8]
```diff
@@ -267,9 +271,11 @@
             "      </Item>\n" +
             "      <vmw:ExtraConfig vmw:key=\"monitor_control.pseudo_perfctr\" vmw:value=\"TRUE\"></vmw:ExtraConfig>\n" +
             "    </VirtualHardwareSection>\n" +
-            "</VirtualSystem>";
+            "</VirtualSystem>" +
+            "</Envelope>";
 
     private String eulaSections =
+            "<Envelope xmlns=\"http://schemas.dmtf.org/ovf/envelope/1\" xmlns:cim=\"http://schemas.dmtf.org/wbem/wscim/1/common\" xmlns:ovf=\"http://schemas.dmtf.org/ovf/envelope/1\" xmlns:rasd=\"http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_ResourceAllocationSettingData\" xmlns:vmw=\"http://www.vmware.com/schema/ovf\" xmlns:vssd=\"http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_VirtualSystemSettingData\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\">" +
```
Comment (DaanHoogland) [inline api/src/test/java/com/cloud/agent/api/storage/OVFHelperTest.java:278]: and here

File: api/src/test/java/com/cloud/agent/api/storage/OVFHelperTest.java [hunk 9]
```diff
@@ -392,9 +398,11 @@
...
             "6. If the Software is subject to Cisco's Smart Licensing program, Cisco will be able to assess if Customer is using the Software within the limits and entitlements paid for by Customer. If the Smart Licensing program is applicable, Customer will be required to enter into a separate terms of service agreement relating to Smart Licensing.</License>\n" +
             "</EulaSection>\n" +
-            "</VirtualSystem>";
+            "</VirtualSystem>" +
+            "</Envelope>";
 
     private String productSectionWithCategories =
+            "<Envelope xmlns=\"http://schemas.dmtf.org/ovf/envelope/1\" xmlns:cim=\"http://schemas.dmtf.org/wbem/wscim/1/common\" xmlns:ovf=\"http://schemas.dmtf.org/ovf/envelope/1\" xmlns:rasd=\"http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_ResourceAllocationSettingData\" xmlns:vmw=\"http://www.vmware.com/schema/ovf\" xmlns:vssd=\"http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_VirtualSystemSettingData\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\">" +
```
Comment (DaanHoogland) [inline api/src/test/java/com/cloud/agent/api/storage/OVFHelperTest.java:405]: and ...

**Review by DaanHoogland — APPROVED** [review]
CLTGM, good work. One suggestion for tests is adding an override on child elements for prefixes and schemas, to see if it can handle those corner cases.

**Review by sureshanaparti — APPROVED** [review]
@nvazquez changes LGTM, please address outstanding comments. Thanks.

**Comment by nvazquez** [comment]
@DaanHoogland @sureshanaparti done, thanks
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test centos7 vmware-67u3 keepEnv

**Comment by DaanHoogland** [comment]
@nvazquez @sureshanaparti is further testing needed on this?
Can you describe your testing please @nvazquez ?

**Comment by nvazquez** [comment]
Hi @DaanHoogland, yes it needs testing. I have tested the PR registering different valid templates - we were aware of some templates with prefixes so used them to test

**Comment by DaanHoogland** [comment]
@nvazquez can you please describe the testing in this PR?

**Comment by nvazquez** [comment]
@DaanHoogland I already did, but can extend it a bit more:

- Register template (multiple times, including templates containing prefixes)
- Check each template registration is successful and deploy from them

**Comment by DaanHoogland** [comment]
> @DaanHoogland I already did, but can extend it a bit more:
> 
>     * Register template (multiple times, including templates containing prefixes)
> 
>     * Check each template registration is successful and deploy from them

I added this to the PR description

**Review by borisstoyanov — APPROVED** [review]
LGTM, manually registered templates and confirmed the fix

**Comment by nvazquez** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by sureshanaparti** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by nvazquez** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by sureshanaparti** [comment]
> Trillian test result (tid-2991) Environment: vmware-67u3 (x2), Advanced Networking with Mgmt server 7 Total time taken: 37246 seconds Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr5825-t2991-vmware-67u3.zip Smoke tests completed. 91 look OK, 1 have errors Only failed tests results shown below:
> 
> Test	Result	Time (s)	Test File
> test_01_nic	`Error`	182.42	test_nic.py

the test failure ^^^ is not related to this PR changes
